File: server/player/mplayer.py
import pexpect
from collections import namedtuple

class MPlayer:
  def __init__(self):
    self._proc = pexpect.spawn('mplayer', timeout=5,
      args=[
        '-slave',
        '-idle',
        '-quiet',
        '-vo', 'null',
        '-softvol',
        '-nolirc'])
    self._proc.flush()

    Property = namedtuple('Property', 'gettable settable type')
    self._prop = {
        'length'    : Property(True,    False,  'time'),
        'time_pos'  : Property(True,    True,   'time'),
        'pause'     : Property(True,    False,  'flag'),
        'filename'  : Property(True,    False,  'string'),
        'volume'    : Property(True,    True,   'float')
    }

  def _write(self, data, expect = None):
    print 'STDIN:', data
    self._proc.write(data + '\n')
    re = None
    if expect is not None:
      re = self._proc.expect(expect)
    self._proc.flush()
    return re

  def _getProperty(self, name):
    if name in self._prop and self._prop[name].gettable:
      prop = self._prop[name]

      base_expected = [#'ANS_ERROR=PROPERTY_UNAVAILABLE',
          'Failed to get value of property \'' + name + '\'.',
          pexpect.TIMEOUT]
      if prop.type == 'time':
        exp = ['ANS_' + name + '=(\d+\.\d+)']
      elif prop.type == 'float':
        exp = ['ANS_' + name + '=(\d+\.\d+)']
      elif prop.type == 'int':
        exp = ['ANS_' + name + '=(\d+)']
      elif prop.type == 'string':
        exp = ['ANS_' + name + '=([-.0-9a-zA-Z_ ]+)']
      elif prop.type == 'flag':
        exp = ['ANS_' + name + '=yes', 'ANS_' + name + '=no']
      exp = exp + base_expected

      index = self._write('pausing_keep_force get_property ' + name, exp)
      if prop.type == 'flag':
        if index == 0:
          return True
        if index == 1:
          return False
      if index == 0:
        data = self._proc.match.group(1)
        if prop.type == 'float':
          return float(data)
        if prop.type == 'time':
          return float(data)
        if prop.type == 'int':
          return int(data)
        if prop.type == 'string':
          return data
    return None

  def _setProperty(self, name, value):
    if name in self._prop and self._prop[name].settable:
      prop = self._prop[name]
      self._write('pausing_keep_force set_property ' + name + ' ' + str(value))

  def kill(self, signal):
    # TODO
    pass

  def quit(self, code = 0):
    self._write('quit ' + str(code))
    self._proc.wait()

  def loadfile(self, filename, append = '0'):
    self._write('loadfile "' + filename + '" ' + str(append),
        ['Starting playback...',
          'Failed to open ' + filename + '.',
          pexpect.TIMEOUT])

  def forcePause(self):
    if self.isPaused() is False:
      self._write('pause')

  def forceUnpause(self):
    if self.isPaused() is True:
      self._write('pause')

  def songOver(self):
    return self.getFilename() is None

  def getFilename(self):
    return self._getProperty('filename')

  def getVolume(self):
    return self._getProperty('volume')

  def getTimePos(self):
    return self._getProperty('time_pos')

  def getLength(self):
    return self._getProperty('length')

  def isPaused(self):
    return self._getProperty('pause')

  def setTimePos(self, value):
    self._setProperty('time_pos', value)

  def setVolume(self, value):
    self._setProperty('volume', value)

if __name__ == '__main__':
  mplayer = MPlayer()
  print mplayer.getFilename()
  print mplayer.getTimePos()
  print mplayer.getLength()
  print mplayer.getVolume()
  print mplayer.isPaused()
  mplayer.loadfile('../../media/MGMT/Electric Feel - Justice remix.mp3')
  mplayer.setTimePos(10)
  mplayer.setVolume(50)
  mplayer.forcePause()
  print mplayer.getFilename()
  print mplayer.getTimePos()
  print mplayer.getLength()
  print mplayer.getVolume()
  print mplayer.isPaused()
  mplayer.forceUnpause()
  print mplayer.isPaused()
  mplayer.quit()

# vim: ft=python et sw=2 sts=2

File: server/player/mplayer.js
var spawn = require('child_process').spawn;

var Request = function(input, outWaiting, errWaiting, callback, timeout) {
  this._input = input;
  this._outWaiting = outWaiting;
  this._errWaiting = errWaiting;

  this._onFinish = callback;

  this._checkedOut = false;
  this._checkedErr = false;

  var baseTimeout = 30000;
  this._timeout = timeout || baseTimeout;
  this._beginTime = new Date().getTime();
};

Request.prototype.ask = function() {
  return this._input;
};

Request.prototype.checked = function() {
  return this._checkedOut || this._checkedErr
    || (this._outWaiting === null && this._errWaiting == null);
};

Request.prototype.timeout = function() {
  return new Date().getTime() - this._beginTime > this._timeout;
};

Request.prototype.checkOut = function(data) {
  if (this._checkedOut) { return this._checkedOut; }

  var re = false;
  if (this._outWaiting === null) {
    re = true;
    if (this._errWaiting !== null) { re = false; }
  } else if (this._outWaiting === undefined) {
    re = false;
  } else {
    var checkWord = function(word) {
      return data.length >= word.length
        && data.slice(0, word.length) == word;
    };

    if (typeof this._outWaiting == 'array') {
      for (var i = 0; i < this._outWaiting.length; i++) {
        var word = this._outWaiting[i];
        re = checkWord(word);
        if (re) { break; }
      }
    } else {
      re = checkWord(this._outWaiting);
    }
  }

  if (re === true && this._onFinish !== undefined) {
    this._onFinish(data);
    this._onFinish = undefined;
  }
  this._checkedOut = re;
  return re;
};

Request.prototype.checkErr = function(data) {
  if (this._checkedErr)
    return this._checkedErr;

  var re = false;
  if (this._errWaiting === null) {
    re = true;
    if (this._outWaiting !== null)
      re = false;
  } else if (this._errWaiting === undefined) {
    re = false;
  } else {
    var checkWord = function(word) {
      if (data.length >= word.length
          && data.slice(0, word.length) == word)
        return true;
      return false;
    };

    if (Array.isArray(this._errWaiting)) {
      for (var i = 0; i < this._errWaiting.length; i++) {
        var word = this._errWaiting[i];
        re = checkWord(word);
        if (re)
          break;
      }
    } else {
      re = checkWord(this._errWaiting);
    }
  }

  if (re === true && this._onFinish !== undefined) {
    this._onFinish(data);
    this._onFinish = undefined;
  }
  this._checkedErr = re;
  return re;
};

var Mplayer = module.exports = function() {
  // 3 input/output fifo
  this._inStack = [];
  this._outStack = [];
  this._errStack = [];

  // Current request
  this._request = null;

  // Subproc of mplayer
  this._procOpened = false;
  this._procOpening = false;
  this._process = null;

  // Load the list of properties
  this._properties = {
    'pause':    { type: 'flag' },
    'length':   { type: 'time' },
    'volume':   { type: 'float' },
    'time_pos': { type: 'time' },
    'filename': { type: 'string' }
  };

  // Flush, listen and update sometimes
  var self = this, timeToUpdate = 200;
  setInterval(function() { self.flush(); }, timeToUpdate);
  setInterval(function() { self.listen(); }, timeToUpdate);
};

Mplayer.prototype._log = function() {
  var args = [].slice.call(arguments, 0);
  args.unshift('[Mplayer]');
  console.log.apply(console, args);
};

Mplayer.prototype.update = function(callback) {
  var self = this;
  self._getProperty('filename', function() {
    self._getProperty('volume', function() {
      self._getProperty('length', function() {
        self._getProperty('time_pos', function() {
          self._getProperty('pause', function() {
            callback();
          });
        });
      });
    });
  });
};

Mplayer.prototype.start = function() {
  if (this._procOpened || this._procOpening) { return; }

  this._procOpening = true;
  this._process = spawn('mplayer', ['-slave', '-idle', '-quiet', '-softvol', '-nolirc', '-vo', 'NULL']);

  // Active the communication with the process after a little time
  // TODO find another way to detect if it's on
  var self = this;
  var timer = setTimeout(function() {
    self._procOpened = true;
    self._procOpening = false;
  }, this._procWaitingTime);

  this._process.stdout.on('data', function(data) {
    var lines = (new String(data)).split('\n');
    self._log('STDOUT', lines);
    for (var i = 0; i < lines.length; i++) {
        self._outStack.push(lines[i]);
    }
    self.listen();
    self.flush();
  });

  this._process.stderr.on('data', function(data) {
    var lines = (new String(data)).split('\n');
    self._log('STDERR', lines);
    for (var i = 0; i < lines.length; i++) {
      self._errStack.push(lines[i]);
    }
    self.listen();
    self.flush();
  });

  this._process.on('exit', function(code, signal) {
    clearInterval(timer);
    self._procOpened = false;
    self._procOpening = false;
    self._process = null;

    // Restart mplayer
    self.start();
  });
};

Mplayer.prototype.kill = function(signal) {
  if (signal === undefined) { signal = 'SIGTERM'; }
  if (this._process) { this._process.kill(signal); }
};

Mplayer.prototype.flush = function() {
  if (!this._procOpened) { return; }

  while (this._request === null && this._inStack.length > 0) {
    this._request = this._inStack.shift();
    this._write(this._request.ask());
  }
};

Mplayer.prototype.listen = function() {
  if (!this._procOpened || this._request === null) { return; }

  // Check output fifo
  while (this._outStack.length > 0) {
    var out = this._outStack.shift();
    this._request.checkOut(out);
  };

  // Check error fifo
  while (this._errStack.length > 0) {
    var err = this._errStack.shift();
    this._request.checkErr(err);
  };

  if (this._request.checked()) {
    this._request = null;
  } else if (this._request.timeout()) {
    this._inStack.unshift(new Request(this._request._input,
          this._request._outWaiting,
          this._request._errWaiting,
          this._request._onFinish,
          this._request._timeout));
    this._request = null;
  }

  // Flush both output fifo
  for (; this._outStack.length; this._outStack.shift());
  for (; this._errStack.length; this._errStack.shift());
};

Mplayer.prototype._write = function(data) {
  if (!this._procOpened) { return; }
  this._log('STDIN', data);
  this._process.stdin.write(data + '\n');
};

Mplayer.prototype._getProperty = function(prop, callback) {
  // TODO more secured
  if (this._properties[prop] === undefined) { return; }

  var self = this, word = 'ANS_' + prop + '=';
  this._inStack.push(new Request('pausing_keep_force get_property ' + prop,
        word, "Failed to get value of property '" + prop + "'",
        function(data) {
          var type = self._properties[prop].type;
          self._properties[prop].value = undefined;
          if (data.length > word.length && data.slice(0, word.length) == word) {
              if      (type == 'float')   { self._properties[prop].value = parseFloat(data.slice(word.length)); }
              else if (type == 'int')     { self._properties[prop].value = parseInt(data.slice(word.length)); }
              else if (type == 'flag')    { self._properties[prop].value = data.slice(word.length) == 'yes'; }
              else if (type == 'string')  { self._properties[prop].value = data.slice(word.length); }
              else if (type == 'pos')     { self._properties[prop].value = parseInt(data.slice(word.length)); }
              else if (type == 'time')    { self._properties[prop].value = parseInt(data.slice(word.length)); }
            }
          if (callback) { callback(self._properties[prop].value); }
        }, 5000));
  this.flush();
};

Mplayer.prototype._setProperty = function(prop, value, callback) {
  // TODO more secured
  if (this._properties[prop] === undefined) { return; }

  var self = this;
  this._inStack.push(new Request(
        'pausing_keep_force set_property ' + prop + ' ' + value,
        null, null,
        function() {
          //self._properties[prop].value = value;
          if (callback) { callback(); }
        }));
  this.flush();
  this.listen();
};

Mplayer.prototype.quit = function(code, callback) {
  if (code === undefined) { code = 0; }
  this._inStack.push(new Request('quit', null, null, callback));
};

Mplayer.prototype.loadFile = function(filename, append, callback) {
  this._inStack.push(new Request(
        'loadfile "' + filename + '" ' + append,
        'Starting playback...',
        'Failed to open ' + filename + '.',
        callback));
};

Mplayer.prototype.forcePause = function(callback) {
  if (!this._properties['pause'].value) {
    this._inStack.push(new Request('pause', null, null, callback));
  }
};

Mplayer.prototype.forceUnpause = function(callback) {
  if (this._properties['pause'].value) {
    this._inStack.push(new Request('pause', null, null, callback));
  }
};

Mplayer.prototype.songOver = function() {
  return (!isNaN(this._properties['time_pos'].value)
      && !isNaN(this._properties['length'].value)
      && this._properties['time_pos'].value >= this._properties['length'].value)
    || this._properties['filename'].value === undefined;
};

Mplayer.prototype.getFilename   = function() { return this._properties['filename'].value; };
Mplayer.prototype.getVolume     = function() { return this._properties['volume'].value; };
Mplayer.prototype.getTimePos    = function() { return this._properties['time_pos'].value; };
Mplayer.prototype.getLength     = function() { return this._properties['length'].value; };
Mplayer.prototype.getPause      = function() { return this._properties['pause'].value; };
Mplayer.prototype.setTimePos    = function(callback) { this._setProperty('time_pos', callback); };
Mplayer.prototype.setVolume     = function(callback) { this._setProperty('volume', callback); };

// vim: ft=javascript et sw=2 sts=2
